### PR TITLE
Add support for other template engines

### PR DESCRIPTION
**Pull request by eduardoboucas**
## 1. Overview

This is a work-in-progress PR to share the progress we're making on adding support for multiple template engines to Web, as per #103. Similarly to how API handles multiple data stores, Web will not ship with any template engine by default, and these would be loaded in as modules.

These modules will implement a standardised API used to communicate with the core. So far, I've created [one for Dust](https://github.com/dadi/web-dustjs) and another [one for Pug](https://github.com/dadi/web-pugjs), with the latter being still in its early stages. I'll put together a repository with a sample interface to make it easy for people to add support for their template engine of choice.

### 2.1. Moving partials directory

As described in #163, we're removing the concept of partials (which is a bit specific to Dust), and instead treating everything as templates. There is still the distinction between templates that create routes (pages) from templates that don't (and are therefore only useful when included from other templates), and this happens with the JSON schema file.

For this reason, to get your Web project ready for this change you'll just need to move your `partials/` directory inside `pages/`.

```
cd my-project
mv workspace/partials workspace/pages
```

### 2.2. Including template engine modules

After pulling this branch, you'll need to choose a template engine to use and include that as a dependency. This is how a `package.json` would look like:

```json
{
  "name": "web-test",
  "version": "1.0.0",
  "main": "index.js",
  "dependencies": {
    "@dadi/web": "dadi/web#feature/template-engines",
    "@dadi/web-dustjs": "dadi/web-dustjs",
    "@dadi/web-pugjs": "dadi/web-pugjs"
  }
}
```

When starting the app, you'll need to feed it the modules for the template engines you wish to use.

```js
require('@dadi/web')({
  engines: [
    require('@dadi/web-dustjs'),
    require('@dadi/web-pugjs')
  ]
})
```

### 2.3. Update external references to template engine core

If your project is using custom helpers, you might need to change the way they access the Dust engine. Because Web was in control of the template engines, it previously had a `Dust` export, which has now been removed. Since the modules will be loaded on the end project, you can access the module directly.

```js
// Before
var dust = require('@dadi/web').Dust
require('@dadi/dustjs-helpers')(dust.getEngine())

// After
var dust = require('dustjs-linkedin')
require('@dadi/dustjs-helpers')(dust)
```

### 2.4. Updating Dust config block in config

We currently have a `dust` block in the app config, with configuration parameters that are specific to Dust.js. This is important, as different engines will require different sets of configuration data. To this effect, I've added a [config block](https://github.com/dadi/web-dustjs/blob/master/index.js#L2-L53) to the metadata object of the engine interface.

The only change required is to **move the `dust` config block** to `engines.dust`. For example:

```json
  "engines": {
    "dust": {
      "cache": true,
      "debug": true,
      "debugLevel": "DEBUG",
      "whitespace": true,
      "paths": {
        "helpers": "workspace/utils/helpers"
      }
    }
  }
```

After following these steps, any existing project using Web should work as expected.

## 3. Adding a page with another engine

Multiple template engines can be used simultaneously with an instance of Web. As an example, and to also show the support for pages in sub-directories suggested by @jimlambie in #103, try the following:

1. Create a `pages/subdir` directory

1. Inside, add a `deep.json` file:

   ```json
   {
     "page": {
       "name": "deep",
       "description": "A view with Pug.",
       "language": "en"
     }
   }
   ```

1. Add a `deep.pug` file:

   ```pug
   doctype html
   html(lang="en")
     head
       title= pageTitle
     body
       h1 Pug - node template engine
       include deeper/foo.pug
   ```

1. Add a `deeper` sub-directory

1. Add a `foo.pug` file inside that subdirectory

   ```html
   <h3>We must go deeper!</h3>
   ```

1. Access that page on `http://localhost/subdir/deep`.

## 4. To do

- [x] Create sample engine interface
- [x] Fix core unit tests
- [x] Add unit/acceptance tests for template store

### CODE REVIEW

**Comment by eduardoboucas** [comment]
Here's a first stab at the sample engine interface: https://github.com/dadi/web-sample-engine

**Comment by eduardoboucas** [comment]
I have removed the `dust` block from the global app config, which means the Dust.js logic is now 100% decoupled from core. I've added a new section (2.4) to the PR description, explaining what was done and what is needed to migrate.

File: test/unit/view.js [hunk 1]
```diff
@@ -1,317 +1,136 @@
...
-    v.render(function (err, result) {
+    v.setData(data)
+    v.render(function(err, result) {
+      var expected = "Sir Moe\nSir Larry\nSir Curly\n"
       result.should.eql(expected)
       done()
     })
   })
-
-  it('should have access to custom dust helpers when calling `render()`', function (done) {
```
Comment (eduardoboucas) [inline test/unit/view.js]: The tests removed from this file have been added to the [Dust interface module](https://github.com/dadi/web-dustjs).

**Comment by abovebored** [comment]
In the `engines` config block, are `engine.dust.paths.helpers` and `engine.dust.paths.filters` overriding the `paths.helpers` & `paths.filters` or do they need to be there instead of?

**Comment by eduardoboucas** [comment]
> In the engines config block, are engine.dust.paths.helpers and engine.dust.paths.filters overriding the paths.helpers & paths.filters or do they need to be there instead of?

Because `helpers` and `filters` are something specific to Dust, I've removed these parameters from the global config object. Instead, they should go in the Dust-specific config block, as `engine.dust.paths.helpers` and `engine.dust.paths.filters`, respectively.

> I like the idea of Web shipping without a default engine, however this breaks our nice '1 line' npm i @dadi/web flow, I'm wondering if it should ship with Dust.js as the default for ease of use. Alternatively, similar to how we copy the default workspace, could add a postinstall script that writes to the package.json for fresh installs? Or some sort of command line setup wizard? "Do you want a default workspace? (y/n). Which template engine do you want to use?" etc

Sure, that makes sense.

**Comment by eduardoboucas** [comment]
Added a `engine` property to the page schema that allows developers to manually specify which template engine should be used for that particular page, overriding whatever engine would be chosen by default based on the file extensions.

For example, the following schema defines a page using Dust.js, even though the template file is named `404.pug`:

```json
{
  "page": {
    "key": "404"
  },
  "engine": "dust"
}
```

**Pull request merged**